stage0/gc.go

```diff
@@ -65,34 +67,86 @@ type mount struct {
 	id         int
 	parentId   int
 	mountPoint string
+	mode       string
 }
 
 type mounts []*mount
 
-// Less ensures that mounts are sorted in an order we can unmount; child before
-// parent. This works by pushing all child mounts up to the front. Parent
-// mounts are halted once they encounter a child. The actual sequential order
-// of the mounts doesn't matter; only that children are in front of parents.
-func (m mounts) Less(i, j int) bool { return m[i].id != m[j].parentId }
-func (m mounts) Len() int           { return len(m) }
-func (m mounts) Swap(i, j int)      { m[i], m[j] = m[j], m[i] }
+// Less ensures that mounts are sorted in an order we can unmount; descendant before ancestor.
+// The requirement of transitivity for Less has to be fulfilled otherwise the sort algorithm will fail.
+// Transitivity is not necessary for our use-case though, so we have to declare extra conditions.
+func (m mounts) getAncestors(i int) map[int]struct{} {
+	ancestorMap := make(map[int]struct{}, 0)
+	current := m[i]
+	found := true
+	for found {
+		for l, mnt := range m {
+			if mnt.id == current.parentId {
+				ancestorMap[mnt.id] = struct{}{}
+				current = mnt
+				break
+			}
+			if l == len(m)-1 {
+				found = false
+				break
+			}
+		}
+	}
+	return ancestorMap
+}
+func (m mounts) Less(i, j int) bool {
+	iAncestors := m.getAncestors(i)
+	jAncestors := m.getAncestors(j)
+	_, jIsAncestor := iAncestors[m[j].id]
+	_, iIsAncestor := jAncestors[m[i].id]
+	result := jIsAncestor || !iIsAncestor
+	result = result || (len(iAncestors) > len(jAncestors))
+	result = !strings.HasPrefix(m[j].mountPoint, m[i].mountPoint) && result
```

Comment:
Maybe `result = result && !strings.HasPrefix(...)`. So the `&&` can short-circuit on `result`.